### PR TITLE
Added command line options to configure code generation

### PR DESCRIPTION
**Pull request by luisggpina**
I couldn't find any good Haskell library for format strings, I can update this PR if you recommend one.  I ended up implementing my own poor-man's format strings.

### CODE REVIEW

**Comment by daniel-grumberg** [comment]
Have you checked out http://hackage.haskell.org/package/text-format-simple ? Seems to do what you want it to do without being able to name the things to format. But I think that keeping the one you wrote is fine, although I think it deserves its own module as it is not strictly related to code generation. I will properly review this tomorrow, though.

File: src/Main.hs [hunk 1]
```diff
@@ -1,30 +1,88 @@
...
+       let (hFile, cFile) = genCode (file_name opts) (headers $ includes opts) (applyPrefixes opts) ir
+       writeFile (out ++ (file_name opts) ++ ".h") hFile
+       writeFile (out ++ (file_name opts) ++ ".c") cFile
   where report err = hPutStrLn stderr ("Error : " ++ show err) >> exitFailure
-        reportList es 
-          = (sequence_ . intersperse (hPutStrLn stderr "") . map (hPutStrLn stderr . show)) es
+        reportList' f es
+          = (sequence_ . intersperse (hPutStrLn stderr "") . map (hPutStrLn stderr . f)) es
             >> exitFailure
+        reportListStr = reportList' id
```
Comment (daniel-grumberg) [inline src/Main.hs]: Nice code reuse ! This a neat refactoring

File: src/Main.hs [hunk 2]
```diff
@@ -1,30 +1,88 @@
...
+
+data MainOptions = MainOptions
+  { fpref       :: String
+  , dpref       :: String
+  , file_name   :: String
+  , includes    :: Includes
+  }
+
+instance Options MainOptions where
+    defineOptions = pure MainOptions
```
Comment (luisggpina) [inline src/Main.hs]: Ok, updated this in 117fb731c8c334e52c920a434f33a4553d2f349f

Using `<$>` on the first line and then `<*>` on the others somehow doesn't feel so clean as using `<*>` everywhere, but I should respect the existing style.

**Comment by luisggpina** [comment]
Please, let me know before you merge so I can rebase all the minor commits to clean the history

File: src/CodeGen.hs [hunk 3]
```diff
@@ -37,15 +101,14 @@ makeImpl (incs, (start, finish), rules)
     declareFps <$$>
     initFunc rules <$$>
     apiImpl (start,finish) <> text "\n" 
-  where libs = text "#include \"divergence.h\"" <$$>
-               text "#include <stdlib.h>" <$$>
-               text "#include <string.h>" <$$>
-               text "#include <asm/unistd.h>" 
+  where libs = text ("#include \"" ++ filename ++ ".h\"") <$$>
+               (foldl (flip $ (<$$>) . text) empty headers)
```
Comment (luisggpina) [inline src/CodeGen.hs]: Updated in 58bdfbf66855ba5089ca267e970e6b0812d67934

**Comment by luisggpina** [comment]
Extracted `Names` module in commit 117fb731c8c334e52c920a434f33a4553d2f349f

The `CodeGen` and `Names` modues are still coupled by the strings that get replaced (e.g. `struct %syscall%`).  I wonder if the `Names` should export the tokens that will be replaced, instead of them being hard-coded in `CodeGen`.  Perhaps you can do a more elegant solution with the `Reader` monad (I'm expecting it to be similar to using a `Holey` monad)

**Comment by daniel-grumberg** [comment]
ok if you can confirm me this does not break `make check` then I will merge

**Comment by daniel-grumberg** [comment]
Well, the `Reader` monad solution would be different to the `Holey` monad solution, it would have just carried around the dictionary and used that. Nevertheless, I was unaware of the `Holey` Monad but it from the little I have seen it seems more appropriate so I will look into it. I will have a hopefully cleaner monadic implementation by Thursday evening, so we can compare then.

**Comment by luisggpina** [comment]
All tests in `make check` are passing, I've rewritten the history to look cleaner without those minor commits.

Merge at your discretion

**Pull request merged**